### PR TITLE
feat(LocalVersion):Allow to have custom embedding models for ingestion and qdrant service

### PR DESCRIPTION
**Pull request by jccscopeo**
Allow the ingestion to use custom embedding if provided in credentials, as well as the custom embedding for the qdrant service if provider of the data source is matching a custom embedding

### CODE REVIEW

File: engine/llm_services/llm_service.py [hunk 1]
```diff
@@ -58,10 +58,16 @@ def __init__(
...
+        embedding_size: int = 3072,
         api_key: Optional[str] = None,
         base_url: Optional[str] = None,
     ):
         super().__init__(trace_manager, provider, model_name, api_key, base_url)
+        self._embedding_size = embedding_size
+
+    def get_embedding_size(self) -> int:
+        """Return the size of the embeddings for the model."""
+        return self._embedding_size
```
Comment (dianascopeo) [inline engine/llm_services/llm_service.py:70]: why function?

File: ingestion_script/ingest_folder_source.py [hunk 2]
```diff
@@ -32,11 +32,23 @@
...
-)
+if settings.custom_embedding_models is not None:
+    # TODO: add the selection at the user level
+    embedding_model_provider = list(settings.custom_embedding_models.keys())[0]
+    embedding_base_url = settings.custom_embedding_models[embedding_model_provider]["base_url"]
+    embedding_api_key = settings.custom_embedding_models[embedding_model_provider]["api_key"]
+    embedding_model_name = settings.custom_embedding_models[embedding_model_provider]["model_name"][0]
+    embedding_model_embedding_size = settings.custom_embedding_models[embedding_model_provider]["embedding_size"][
+        embedding_model_name
+    ]
```
Comment (dianascopeo) [inline ingestion_script/ingest_folder_source.py:43]: same here why list[0]?

**Review by dianascopeo — CHANGES_REQUESTED** [review]

File: engine/llm_services/llm_service.py (hunk 1 above)
Comment (jccscopeo) [reply engine/llm_services/llm_service.py:70]: it is a property also but it is to allow access to the information of the embedding size

File: ingestion_script/ingest_folder_source.py (hunk 2 above)
Comment (jccscopeo) [reply ingestion_script/ingest_folder_source.py:43]: because for now we don t allow user to choose a model for embedding during ingestion. So until we have the front part we will force to use the first embedding model

Comment (dianascopeo) [reply ingestion_script/ingest_folder_source.py]: i dont know if this is a good idea

File: credentials.env.example [hunk 3]
```diff
@@ -102,4 +102,9 @@ PGADMIN_PORT=5050
...
 # Custom LLM configuration
 CUSTOM_LLM_MODELS={"provider_name": {"model_name":["my-llm-1-model-name", "my-llm-2-model-name"], "base_url": "https://llm1.example.com/v1/", "api_key": "my_api_key}}
-CUSTOM_EMBEDDING_MODELS={"provider_name": {"model_name":["my-embed-1-model-name", ""my-embed-2-model-name""], "base_url": "https://embed1.example.com/v1/", "api_key": "my_api_key}}
\ No newline at end of file
+CUSTOM_EMBEDDING_MODELS={"provider_name": {"model_name":["my-embed-1-model-name", ""my-embed-2-model-name""], "base_url": "https://embed1.example.com/v1/", "api_key": "my_api_key}}
+
+# INGESTION LOCAL PARAMETERS
+ZOOM_INGESTION=1.0
+NUMBER_OF_IMAGES_TO_DETERMINE_TYPE_OF_DOCUMENT=2
+INGESTION_WITH_OPEN_SOURCE_LLM=False
```
Comment (natacha-scopeo) [inline credentials.env.example:110]: ingestion_with_open_source shouldn't be ENFORCE_PAGE_BY_PAGE_INGESTION ?

File: ingestion_script/ingest_folder_source.py [hunk 4]
```diff
@@ -22,21 +22,55 @@
...
     trace_manager=TraceManager(project_name="ingestion"),
 )
-OPENAI_COMPLETION_SERVICE = VisionService(
+LLM_SERVICE = VisionService(
     provider="openai",
     model_name="gpt-4.1-mini",
     trace_manager=TraceManager(project_name="ingestion"),
 )
+if settings.custom_llm_models is not None:
+    if len(settings.custom_llm_models) > 0:
```
[the remark this answers is not in the document]
Comment (jccscopeo) [reply ingestion_script/ingest_folder_source.py:36]: Not sure I understand your request

File: ingestion_script/ingest_folder_source.py [hunk 5]
```diff
@@ -22,21 +22,55 @@
...
+    # TODO: add the selection at the user level
+    if len(settings.custom_embedding_models) > 0:
+        embedding_model_provider = list(settings.custom_embedding_models.keys())[0]
+        embedding_base_url = settings.custom_embedding_models[embedding_model_provider]["base_url"]
+        embedding_api_key = settings.custom_embedding_models[embedding_model_provider]["api_key"]
+        embedding_model_name = settings.custom_embedding_models[embedding_model_provider]["model_name"][0]
+        embedding_model_embedding_size = settings.custom_embedding_models[embedding_model_provider]["embedding_size"][
+            embedding_model_name
+        ]
+        EMBEDDING_SERVICE = EmbeddingService(
```
[the remark this answers is not in the document]
Comment (jccscopeo) [reply ingestion_script/ingest_folder_source.py:66]: I can put if else if you want

File: credentials.env.example (hunk 3 above)
Comment (jccscopeo) [reply credentials.env.example:110]: You are right I will change it

File: ada_backend/services/entity_factory.py [hunk 6]
```diff
@@ -361,12 +362,31 @@ def processor(params: dict, constructor_params: dict[str, Any]) -> dict:
...
+                    embedding_model_embedding_size = settings.custom_embedding_models[provider]["embedding_size"][
+                        model_name
+                    ]
+                    embedding_service = EmbeddingService(
+                        trace_manager=get_trace_manager(),
+                        api_key=embedding_api_key,
+                        provider=provider,
+                        model_name=model_name,
+                        base_url=embedding_base_url,
+                        embedding_size=embedding_model_embedding_size,
```
[the remark this answers is not in the document]
Comment (jccscopeo) [reply ada_backend/services/entity_factory.py:380]: what do you mean?

File: engine/llm_services/llm_service.py [hunk 7]
```diff
@@ -423,6 +444,17 @@ def get_image_description(
...
 
                 client = openai.OpenAI(api_key=self._api_key, base_url=settings.GOOGLE_BASE_URL)
+            case _:
+                import openai
+
+                if self._api_key is None:
+                    raise ValueError(f"API key is not set for the provider {self._provider}.")
+                if self._base_url is None:
+                    raise ValueError(f"Base URL is not set for the provider {self._provider}.")
+                client = openai.OpenAI(api_key=self._api_key, base_url=self._base_url)
```
[the remark this answers is not in the document]
Comment (jccscopeo) [reply engine/llm_services/llm_service.py:454]: ok

File: data_ingestion/document/pdf_vision_ingestion.py [hunk 8]
```diff
@@ -240,7 +243,7 @@ def create_chunks_from_document(
         prompt=PROMPT_DETERMINE_FILE_TYPE,
         google_llm_service=google_llm_service,
         openai_llm_service=openai_llm_service,
-        image_content_list=images_content_list,
+        image_content_list=images_content_list[:number_of_images_to_determine_type_of_document],
```
[the remark this answers is not in the document]
Comment (dianascopeo) [reply data_ingestion/document/pdf_vision_ingestion.py:246]: for pdf ingetsion we check if it is presentation or normal pdf

File: ingestion_script/ingest_folder_source.py [hunk 9]
```diff
@@ -22,21 +22,57 @@
...
+        temperature=0.0,
+    )
+    LLM_SERVICE = VISION_COMPLETION_SERVICE
+else:
+    VISION_COMPLETION_SERVICE = VisionService(
+        provider="google",
+        model_name="gemini-2.0-flash-exp",
+        trace_manager=TraceManager(project_name="ingestion"),
+    )
+    LLM_SERVICE = VisionService(
```
Comment (dianascopeo) [inline ingestion_script/ingest_folder_source.py:47]: maybe VISION_COMPLETION_SERVICE_OPENAI?

File: ingestion_script/ingest_folder_source.py [hunk 10]
```diff
@@ -22,21 +22,57 @@
...
+        trace_manager=TraceManager(project_name="ingestion"),
+        provider=llm_model_provider,
+        model_name=llm_model_name,
+        api_key=llm_api_key,
+        base_url=llm_base_url,
+        temperature=0.0,
+    )
+    LLM_SERVICE = VISION_COMPLETION_SERVICE
+else:
+    VISION_COMPLETION_SERVICE = VisionService(
```
Comment (dianascopeo) [inline ingestion_script/ingest_folder_source.py:42]: VISION_COMPLETION_SERVICE_GOOGLE?

File: ingestion_script/ingest_folder_source.py [hunk 11]
```diff
@@ -22,21 +22,57 @@
...
+    llm_model_name = settings.custom_llm_models[llm_model_provider]["model_name"][0]
+    VISION_COMPLETION_SERVICE = VisionService(
+        trace_manager=TraceManager(project_name="ingestion"),
+        provider=llm_model_provider,
+        model_name=llm_model_name,
+        api_key=llm_api_key,
+        base_url=llm_base_url,
+        temperature=0.0,
+    )
+    LLM_SERVICE = VISION_COMPLETION_SERVICE
```
Comment (dianascopeo) [inline ingestion_script/ingest_folder_source.py:40]: I did not understand this

File: ingestion_script/ingest_folder_source.py [hunk 12]
```diff
@@ -22,21 +22,57 @@
...
-)
-EMBEDDING_SERVICE = EmbeddingService(
-    provider="openai",
-    model_name="text-embedding-3-large",
-    trace_manager=TraceManager(project_name="ingestion"),
-)
+
+# TODO: add the selection at the user level
+if (settings.custom_llm_models is not None) and (len(settings.custom_llm_models) > 0):
+    llm_model_provider = list(settings.custom_llm_models.keys())[0]
```
[the remark this answers is not in the document]
Comment (jccscopeo) [reply ingestion_script/ingest_folder_source.py:28]: yes for now it is always first by default

File: ingestion_script/ingest_folder_source.py [hunk 13]
```diff
@@ -191,8 +227,8 @@ def _ingest_folder_source(
         )
     try:
         document_chunk_mapping = document_chunking_mapping(
-            vision_ingestion_service=GOOGLE_COMPLETION_SERVICE,
-            llm_service=OPENAI_COMPLETION_SERVICE,
+            vision_ingestion_service=VISION_COMPLETION_SERVICE_GOOGLE,
+            llm_service=VISION_LLM_SERVICE_OPENAI,
```
[the remark this answers is not in the document]
Comment (jccscopeo) [reply ingestion_script/ingest_folder_source.py:231]: It is similar I think. I will rename like that the LLM names.

File: credentials.env.example [hunk 14]
```diff
@@ -102,4 +102,9 @@ PGADMIN_PORT=5050
 
 # Custom LLM configuration
 CUSTOM_LLM_MODELS={"provider_name": {"model_name":["my-llm-1-model-name", "my-llm-2-model-name"], "base_url": "https://llm1.example.com/v1/", "api_key": "my_api_key}}
-CUSTOM_EMBEDDING_MODELS={"provider_name": {"model_name":["my-embed-1-model-name", ""my-embed-2-model-name""], "base_url": "https://embed1.example.com/v1/", "api_key": "my_api_key}}
\ No newline at end of file
+CUSTOM_EMBEDDING_MODELS={"provider_name": {"model_name":["my-embed-1-model-name", ""my-embed-2-model-name""], "base_url": "https://embed1.example.com/v1/", "api_key": "my_api_key}}
+
+# INGESTION LOCAL PARAMETERS
+ZOOM_INGESTION=1.0
```
[the remark this answers is not in the document]
Comment (jccscopeo) [reply credentials.env.example:108]: no problem

File: README.md [hunk 15]
```diff
@@ -309,12 +309,16 @@ You can configure Draft'n run to use your own custom Large Language Model (LLM)
...
+<<<<<<< HEAD
 - **`completion_models`**: List of available completion models
   - `model_name`: The name of your model
   - `function_calling`: Whether the model supports function calling
   - `multimodal`: Whether the model supports multimodal inputs (images, etc.)
   - `constrained_completion_with_pydantic`: Whether the model supports Pydantic-constrained outputs
   - `constrained_completion_with_json_schema`: Whether the model supports JSON schema-constrained outputs
+=======
+CUSTOM_EMBEDDING_MODELS={"provider_name": {"model_name":["my-embed-1-model-name", "my-embed-2-model-name"], "embedding_size" : {"my-embed-1-model-name" : 100, "my-embed-2-model-name" : 200 }, "base_url": "https://embed1.example.com/v1/", "api_key": "my_api_key}}
+>>>>>>> f9fe47d (correct LLM names for ingestion + zoom variable)
```
Comment (dianascopeo) [inline README.md:321]: rebase

File: ada_backend/services/entity_factory.py [hunk 16]
```diff
@@ -361,12 +362,27 @@ def processor(params: dict, constructor_params: dict[str, Any]) -> dict:
...
+            if provider in custom_models:
+                for config_embedding_model in custom_models[provider].get("embedding_models", []):
+                    if config_embedding_model.get("model_name") == model_name:
+                        embedding_service = EmbeddingService(
+                            trace_manager=get_trace_manager(),
+                            api_key=custom_models[provider].get("api_key", None),
+                            provider=provider,
+                            model_name=model_name,
+                            base_url=custom_models[provider].get("base_url", None),
+                            embedding_size=config_embedding_model.get("embedding_size", None),
```
Comment (dianascopeo) [inline ada_backend/services/entity_factory.py:376]: maybe it is better to implement this feature via front ?

File: engine/llm_services/llm_service.py [hunk 17]
```diff
@@ -96,6 +106,8 @@ def embed_text(self, text: str) -> list[float]:
                 return response.data
 
             case _:
+                if self._base_url is None or self._api_key is None:
+                    raise ValueError(f"API key and base URL must be provided for custom provider: {self._provider}")
```
Comment (dianascopeo) [inline engine/llm_services/llm_service.py:110]: why check here as well if you did it in init?

File: engine/llm_services/llm_service.py [hunk 18]
```diff
@@ -397,7 +403,18 @@ def _format_image_content(self, image_content_list: list[bytes]) -> list[dict[st
...
             case _:
-                raise ValueError(f"Invalid provider: {self._provider}")
+
+                import base64
+
+                return [
+                    {
+                        "type": "image_url",
+                        "image_url": {
+                            "url": f"data:image/png;base64,{base64.b64encode(image_content).decode('utf-8')}"
```
[the remark this answers is not in the document]
Comment (dianascopeo) [reply engine/llm_services/llm_service.py:490]: for gemini it is jpeg

File: ingestion_script/utils.py [hunk 19]
```diff
@@ -199,3 +199,45 @@ def upload_source(
...
+                api_key = config_provider.get("api_key")
+                return EmbeddingService(
+                    provider=provider,
+                    model_name=model_name,
+                    trace_manager=TraceManager(project_name="ingestion"),
+                    api_key=api_key,
+                    base_url=base_url,
+                    embedding_size=embedding_size,
+                )
+    return None
```
Comment (dianascopeo) [inline ingestion_script/utils.py:243]: i think you dont need this if you have provider and llm service you can easily get api and base_url it is in init

File: ada_backend/services/entity_factory.py (hunk 16 above)
Comment (jccscopeo) [reply ada_backend/services/entity_factory.py:376]: yes but we wont have it right away hence the workaround

File: engine/llm_services/llm_service.py (hunk 17 above)
Comment (jccscopeo) [reply engine/llm_services/llm_service.py:110]: you are right I will remove it

File: README.md [hunk 20]
```diff
@@ -337,6 +341,26 @@ When you have done all configuration for local models you need to run seed again
 make db-seed
 ```
 
+You can also configure the parameters of the ingestion on the `credentials.env` file:
+
+```env
+PAGE_RESOLUTION_ZOOM=1.0
+NUMBER_OF_IMAGES_TO_DETERMINE_TYPE_OF_DOCUMENT=2
+INGESTION_WITH_OPEN_SOURCE_LLM=False
```
Comment (natacha-scopeo) [inline README.md:349]: you add open source ? is it missing also ENFORCE_PAGE_BY_PAGE_INGESTION ?

File: .github/workflows/.github-ci.yml [hunk 21]
```diff
@@ -123,6 +123,9 @@ jobs:
           echo "S3_SECRET_ACCESS_KEY=${{ secrets.S3_SECRET_ACCESS_KEY }}" >> credentials.env
           echo "S3_BUCKET_NAME=${{ secrets.S3_BUCKET_NAME }}" >> credentials.env
           echo "S3_REGION_NAME=eu-west-3" >> credentials.env
+          echo "PAGE_RESOLUTION_ZOOM=3.0" >> credentials.env
+          echo "NUMBER_OF_IMAGES_TO_DETERMINE_TYPE_OF_DOCUMENT=-1" >> credentials.env
+          echo "ENFORCE_PAGE_BY_PAGE_INGESTION=False" >> credentials.env
```
Comment (natacha-scopeo) [inline .github/workflows/.github-ci.yml:129]: useless if already set default value in settings file

File: credentials.env.example [hunk 22]
```diff
@@ -110,3 +110,7 @@ S3_SECRET_ACCESS_KEY=your_s3_secret_access_key
 S3_BUCKET_NAME=s3-backend
 S3_REGION_NAME=us-east-1
 
+# INGESTION LOCAL PARAMETERS
+PAGE_RESOLUTION_ZOOM=1.0
+NUMBER_OF_IMAGES_TO_DETERMINE_TYPE_OF_DOCUMENT=2
+ENFORCE_PAGE_BY_PAGE_INGESTION=True
```
Comment (natacha-scopeo) [inline credentials.env.example:116]: missing open source value ?

File: data_ingestion/document/pdf_vision_ingestion.py [hunk 23]
```diff
@@ -268,15 +271,17 @@ def create_chunks_from_document(
...
-
-        extracted_table_of_content = _extract_text_from_pages_as_images(
-            prompt=PDF_TABLE_OF_CONTENT_EXTRACTION_PROMPT,
-            google_llm_service=google_llm_service,
-            openai_llm_service=openai_llm_service,
-            image_content_list=images_content_list,
-            response_format=TableOfContent,
-        )
-        if extracted_table_of_content.sections != []:
+        chunking_by_section = False
```
Comment (natacha-scopeo) [inline data_ingestion/document/pdf_vision_ingestion.py:274]: why is the idea of chunking_by_section ?

File: ingestion_script/utils.py [hunk 24]
```diff
@@ -199,3 +199,45 @@ def upload_source(
...
+                    api_key = config_provider.get("api_key")
+                    return VisionService(
+                        trace_manager=TraceManager(project_name="ingestion"),
+                        provider=provider,
+                        model_name=model_name,
+                        api_key=api_key,
+                        base_url=base_url,
+                        temperature=0.0,
+                    )
+    return None
```
Comment (natacha-scopeo) [inline ingestion_script/utils.py:222]: Do we raise error if we find nothing ?

Comment (jccscopeo) [reply ingestion_script/utils.py:217]: hum no, if it does not find anything it will use the default config with geminin and openai and if it is not set up it will throw an error on the ingestion side

File: data_ingestion/document/pdf_vision_ingestion.py (hunk 23 above)
Comment (jccscopeo) [reply data_ingestion/document/pdf_vision_ingestion.py:274]: it was something from Diana. You get the table of contents from all the images and you try to build each section with all the images again in markdown before chunking

**Comment by jccscopeo** [comment]
> How d you decide to use custom llm ?

For now it is by default using the custom llm if it is present, but I can put a boolean instead

File: .github/workflows/.github-ci.yml (hunk 21 above)
Comment (jccscopeo) [reply .github/workflows/.github-ci.yml:129]: no default value in settings

File: README.md (hunk 20 above)
Comment (jccscopeo) [reply README.md:347]: yes my bad I changed INGESTION_WITH_OPEN_SOURCE_LLM by ENFORCE_PAGE_BY_PAGE_INGESTION

File: credentials.env.example [hunk 25]
```diff
@@ -124,3 +125,13 @@ GRAFANA_URL=http://localhost:3000
...
+
+# INGESTION LOCAL PARAMETERS
+PAGE_RESOLUTION_ZOOM=1.0
+NUMBER_OF_IMAGES_TO_DETERMINE_TYPE_OF_DOCUMENT=2
+<<<<<<< HEAD
+ENFORCE_PAGE_BY_PAGE_INGESTION=False
+
+=======
+ENFORCE_PAGE_BY_PAGE_INGESTION=True
+>>>>>>> bcb5159 (correct LLM names for ingestion + zoom variable)
```
Comment (dianascopeo) [inline credentials.env.example:137]: same

File: engine/llm_services/llm_service.py (hunk 17 above)
Comment (dianascopeo) [reply engine/llm_services/llm_service.py:110]: still here

File: engine/llm_services/llm_service.py [hunk 26]
```diff
@@ -63,6 +67,10 @@ def __init__(
                     self._api_key = settings.custom_models.get(self._provider).get("api_key")
                     self._base_url = settings.custom_models.get(self._provider).get("base_url")
                     LOGGER.debug(f"Using custom api key and base url for provider: {self._provider}")
+                    if self._api_key is None or self._base_url is None:
+                        raise ValueError(
+                            f"API key and base URL must be provided for custom provider: {self._provider}"
+                        )
```
Comment (dianascopeo) [inline engine/llm_services/llm_service.py:73]:
```suggestion
                case _:
                    custom_model = settings.custom_models.get(self._provider)
                    if custom_model is None:
                        raise ValueError(f"Provider {self._provider} not found in settings")
                    self._api_key = custom_model.get("api_key")
                    self._base_url = custom_model.get("base_url")
                    LOGGER.debug(f"Using custom api key and base url for provider: {self._provider}")
```

File: engine/llm_services/llm_service.py [hunk 27]
```diff
@@ -452,23 +464,40 @@ def __init__(
     ):
         super().__init__(trace_manager, provider, model_name, api_key, base_url)
         self._temperature = temperature
+        self._image_format = None
+        match self._provider:
+            case "openai":
+                self._image_format = "jpeg"
+            case "google":
+                self._image_format = "jpeg"
```
Comment (dianascopeo) [inline engine/llm_services/llm_service.py:472]:
```suggestion
            case "openai" | "google":
                self._image_format = "jpeg"
```

File: engine/llm_services/llm_service.py [hunk 28]
```diff
@@ -452,23 +464,40 @@ def __init__(
...
+                    if model.get("model_name") == self._model_name:
+                        self._image_format = model.get("image_format", None)
+                        break
+
+                LOGGER.debug(f"Using image format for custom model {self._model_name}: {self._image_format}")
+                if self._image_format is None:
+                    raise ValueError(
+                        f"image format not provided for custom model {self._model_name} "
+                        f"for provider {self._provider}"
+                    )
```
Comment (dianascopeo) [inline engine/llm_services/llm_service.py:487]: why you are doing this check? and not directly get the image_format?

Comment (jccscopeo) [reply engine/llm_services/llm_service.py:487]: not sure I understand

Comment (dianascopeo) [reply engine/llm_services/llm_service.py:487]: why you are doing this?
`                for model in completions_models:
                    if model.get("model_name") == self._model_name:
                        self._image_format = model.get("image_format", None)
                        break`

Comment (dianascopeo) [reply engine/llm_services/llm_service.py:487]: in my opinion if model is multimodal you can use it here if not you will not use

File: data_ingestion/document/pdf_vision_ingestion.py (hunk 23 above)
Comment (jccscopeo) [reply data_ingestion/document/pdf_vision_ingestion.py]: not really it is a default behavior unless a error is detected in the pipeline of ingestion

File: .github/workflows/.github-ci.yml [hunk 29]
```diff
@@ -125,6 +125,10 @@ jobs:
           echo "ENABLE_OBSERVABILITY_STACK=false" >> credentials.env
           echo "SEGMENT_API_KEY=${{ secrets.SEGMENT_API_KEY }}" >> credentials.env
           echo "ENV=${{ secrets.ENV }}" >> credentials.env
+          echo "ZOOM_INGESTION=3.0" >> credentials.env
+          echo "PAGE_RESOLUTION_ZOOM=3.0" >> credentials.env
```
[the remark this answers is not in the document]
Comment (jccscopeo) [reply .github/workflows/.github-ci.yml:129]: I will remove it, it is PAGE_RESOLUTION_ZOOM to keep

File: .github/workflows/.github-ci.yml [hunk 30]
```diff
@@ -125,6 +125,10 @@ jobs:
...
+          echo "NUMBER_OF_IMAGES_TO_DETERMINE_TYPE_OF_DOCUMENT=-1" >> credentials.env
```
[the remark this answers is not in the document]
Comment (jccscopeo) [reply .github/workflows/.github-ci.yml:130]: putting -1 you can use all the images to determine the type of the document. I will put 5 instead

File: engine/llm_services/llm_service.py [hunk 31]
```diff
@@ -90,10 +101,12 @@ def __init__(
         trace_manager: TraceManager,
         provider: str = "openai",
         model_name: str = "text-embedding-3-large",
+        embedding_size: int = 3072,
```
[the remark this answers is not in the document]
Comment (jccscopeo) [reply engine/llm_services/llm_service.py:104]: size of the openai embedding

File: .github/workflows/.github-ci.yml [hunk 32]
```diff
@@ -125,6 +125,10 @@ jobs:
...
+          echo "ENFORCE_PAGE_BY_PAGE_INGESTION=False" >> credentials.env
```
[the remark this answers is not in the document]
Comment (jccscopeo) [reply .github/workflows/.github-ci.yml]: removed

File: data_ingestion/document/pdf_vision_ingestion.py [hunk 33]
```diff
@@ -225,53 +225,111 @@ def _create_chunks_from_markdown(
     return chunks
 
 
+def _create_chunk_from_text(
+    document: FileDocument,
```
[the remark this answers is not in the document]
Comment (natacha-scopeo) [reply data_ingestion/document/pdf_vision_ingestion.py:229]: or me this function is quite useless

**Review by natacha-scopeo — COMMENTED** [review]
It works locally ! Do we need to do make db-seed ?

**Review by dianascopeo — APPROVED** [review]
worked locally tried custom embeddings

File: data_ingestion/document/pdf_vision_ingestion.py [hunk 34]
```diff
@@ -225,53 +225,111 @@ def _create_chunks_from_markdown(
...
+            google_llm_service=google_llm_service,
+            openai_llm_service=openai_llm_service,
+            image_content_list=[img_base64],
+        )
+        chunk = _create_chunk_from_text(document, extracted_text, i + 1)
+        chunks.append(chunk)
+    return chunks
+
+
+def _process_with_table_of_contents(
```
Comment (pabloscopeo) [inline data_ingestion/document/pdf_vision_ingestion.py:269]: maybe `_process_pdf_*` ?

File: engine/llm_services/constrained_output_models.py [hunk 35]
```diff
@@ -58,3 +64,85 @@ class OutputFormatModel(BaseModel):
...
+        elif issubclass(origin, BaseModel):
+            # nested model
+            return generate_fake_instance_from_pydantic(origin)
+        else:
+            return None
+
+    # build data dict
+    data = {name: _fake_value(field.annotation) for name, field in model_cls.model_fields.items()}
+
+    return model_cls(**data)
```
Comment (pabloscopeo) [inline engine/llm_services/constrained_output_models.py:102]: This is an util for testing?

File: engine/llm_services/constrained_output_models.py [hunk 36]
```diff
@@ -58,3 +64,85 @@ class OutputFormatModel(BaseModel):
...
+    Useful when using open source llm with failing constrained output models.
+
+    Args:
+        prompt (str): The initial prompt text.
+        pydantic_model (BaseModel): The Pydantic model instance to be included in the prompt.
+
+    Returns:
+        str: The formatted prompt with the Pydantic model instance.
+    """
+    pydantic_json_instance = generate_fake_instance_from_pydantic(pydantic_model).model_dump_json(indent=2)
```
Comment (pabloscopeo) [inline engine/llm_services/constrained_output_models.py:117]: So this is used to make a json describing a pydantic model for a prompt? If that's the case we should either see if pydantic supports something like that to avoid having to make a fake instance, either make some unit tests for `generate_fake_instance_from_pydantic` and put these functions on another module. (Btw maybe we can say "example" instead of "fake"(?))

File: engine/llm_services/constrained_output_models.py [hunk 37]
```diff
@@ -58,3 +64,85 @@ class OutputFormatModel(BaseModel):
...
+
+def search_for_json_object(json_string: str) -> Optional[str]:
+    index_start_bracket = -1
+    index_end_bracket = -1
+    for i, (char, reverse_char) in enumerate(zip(json_string, json_string[::-1])):
+        if char == "{" and index_start_bracket == -1:
+            index_start_bracket = i
+        if reverse_char == "}" and index_end_bracket == -1:
+            index_end_bracket = len(json_string) - i - 1
+    return json_string[index_start_bracket : index_end_bracket + 1]
```
Comment (pabloscopeo) [inline engine/llm_services/constrained_output_models.py:129]: I would advocate for adding unit tests for this function to be sure it handles edge cases

File: engine/llm_services/constrained_output_models.py [hunk 38]
```diff
@@ -58,3 +64,85 @@ class OutputFormatModel(BaseModel):
...
+
+
+def convert_json_answer_to_pydantic(json_answer: str, pydantic_model: BaseModel) -> BaseModel:
+    """
+    Converts a JSON string answer to a Pydantic model instance.
```
Comment (pabloscopeo) [inline engine/llm_services/constrained_output_models.py:134]: As I understand, this function takes a string that contains a json somewhere, identifies where it is and tries to parse it even if it contains errors right? So maybe the naming could be a bit clearer. Can we come up with something better than "json_answer" for both the function name and the first param? I think "json_answer" it's a bit unclear.

File: engine/llm_services/llm_service.py [hunk 39]
```diff
@@ -570,23 +582,38 @@ def __init__(
...
+
+                LOGGER.debug(f"Using image format for custom model {self._model_name}: {self._image_format}")
+                if self._image_format is None:
+                    raise ValueError(
+                        f"image format not provided for custom model {self._model_name} "
+                        f"for provider {self._provider}"
+                    )
+
+    def _format_image_content(self, image_content_list: list[bytes]) -> list[dict[str, str]]:
+        import base64
```
Comment (pabloscopeo) [inline engine/llm_services/llm_service.py:606]: why is this import here?

File: engine/llm_services/llm_service.py [hunk 40]
```diff
@@ -604,14 +631,21 @@ async def get_image_description_async(
...
-        client = None
         span = get_current_span()
         span.set_attributes({SpanAttributes.LLM_INVOCATION_PARAMETERS: json.dumps({"temperature": self._temperature})})
         match self._provider:
             case "openai" | "google":
                 import openai
 
                 client = openai.AsyncOpenAI(api_key=self._api_key, base_url=self._base_url)
+            case _:
+                import openai
```
Comment (pabloscopeo) [inline engine/llm_services/llm_service.py:642]: why is this import here?

File: ingestion_script/utils.py [hunk 41]
```diff
@@ -216,3 +216,34 @@ def build_combined_sql_filter(
...
+            list_completion_models = config_provider.get("completion_models", [])
+            for model_config in list_completion_models:
+                if model_config.get("multimodal", False):
+                    return VisionService(
+                        trace_manager=TraceManager(project_name="ingestion"),
+                        provider=provider,
+                        model_name=model_config.get("model_name"),
+                        temperature=0.0,
+                    )
+    return None
```
Comment (pabloscopeo) [inline ingestion_script/utils.py:234]: can we return early?
```python
    if not cond:
        return None
    for ...
```

File: ingestion_script/utils.py [hunk 42]
```diff
@@ -216,3 +216,34 @@ def build_combined_sql_filter(
...
+        for provider, config_provider in custom_models.items():
+            list_embeddings_models = config_provider.get("embedding_models", [])
+            for model_config in list_embeddings_models:
+                return EmbeddingService(
+                    provider=provider,
+                    model_name=model_config.get("model_name"),
+                    trace_manager=TraceManager(project_name="ingestion"),
+                    embedding_size=model_config.get("embedding_size"),
+                )
+    return None
```
Comment (pabloscopeo) [inline ingestion_script/utils.py:249]: early return?

**Review by pabloscopeo — COMMENTED** [review]
Could we put test instructions somewhere? PR Description for example or via Slack

**Review by pabloscopeo — COMMENTED** [review]
Tested and working locally! LGTM code wise, I made some comments, I'm ok to merge after treating those.

File: engine/llm_services/constrained_output_models.py (hunk 35 above)
Comment (jccscopeo) [reply engine/llm_services/constrained_output_models.py:102]: no
